### DISCUSSION TITLE
Derived write Atom that accepts values or a callback function.

### DISCUSSION THREAD

**Discussion by joel-daros** [pmndrs/jotai#2512] [Q&A]
I have a basic number atom, and a derived atom with a custom writer function to multiply the atom's value by 2 before writing it:

```typescript
const numberAtom = atom<number>(1);

// Create a derived atom that multiplies the numberAtom value by 2
const doubleAtom = atom(
  (get) => get(numberAtom),
  (get, set) => set(numberAtom, get(numberAtom) * 2)
);
``` 

What do I need to modify in my  `doubleAtom` in a way it accepts a raw value, or a callback function result and keeping to multiply the value by 2.

```typescript
 // OK  - The value is multiplied by 2 and the final result is 4
setDoubleAtom(2); 

// But, sometimes I want to pass a callback function to set my value
setDoubleAtom((currentValue) => currentValue - 1); 
``` 

In this case, if my current value is 4, (4 - 1 = 3), so 3 will be passed as value, and the final atom value should be 6 (double value). 

I would appreciate if you can give an example with the proper Typescript types.

**Comment by dai-shi** [comment]
I think I understand the requirement, but let me confirm: How is `setDoubleAtom` defined? Do you have an implementation idea without TS types?

**Reply by joel-daros** [reply]
Sorry I forgot to mention. `setDoubleAtom()` is just the standard atom setter:

```typescript
const [doubleAtom, setDoubleAtom] = useAtom(doubleAtom);
```

**Reply by dai-shi** [reply]
Is it something like this in JS?
```js
const doubleAtom = atom(
  (get) => get(numberAtom),
  (get, set, fn) => {
    set(numberAtom, get(numberAtom) * 2);
    if (fn) {
      set(numberAtom, fn(get(numberAtom));
    }
  },
);
```

**Reply by joel-daros** [reply]
Almost this. I think this is what I was looking for:

```typescript
// auxiliary function used by doubleAtom
const doubleValue = (value: number) => value * 2;

// Create an atom with an initial value of 1
const numberAtom = atom<number>(1);

// Create a derived atom that multiplies the numberAtom value by 2
const doubleAtom = atom<number, [SetStateAction<number>], void>(
  (get) => get(numberAtom),
  (get, set, update) => {
    const currentValue = get(numberAtom);

    if (typeof update === 'number') {
      set(numberAtom, doubleValue(update));
    }

    if (update instanceof Function) {
      set(numberAtom, doubleValue(update(currentValue)));
    }
  }
);

const [doubleAtom, setDoubleAtom] = useAtom(doubleAtom);

setDoubleAtom(4);  // result is 8
setDoubleAtom((currentValue) => currentValue);  // if current value is 2, the result will be 4.
```

If I set a raw value, `setDoubleAtom(4)` the final result will be always `8`.
If I set a function, and the result of function is `2` the final result will be the double of the result of my function: `4`

Let me know if there is a better way to achieve the same result.
